core/src/net/cidr/v4.rs

```diff
@@ -53,20 +115,49 @@ impl Cidr for Ipv4Cidr {
         self.address
     }
 
+    #[inline]
+    fn contains(&self, address: Self::Addr) -> bool {
+        self.prefix == (self.mask & u32::from(address))
+    }
+
+    #[inline]
+    fn network(&self) -> Self::Addr {
+        Ipv4Addr::from(self.mask & u32::from(self.address))
+    }
+
+    #[inline]
+    fn broadcast(&self) -> Self::Addr {
+        Ipv4Addr::from((self.mask & u32::from(self.address)) | !self.mask)
+    }
+
     #[inline]
     fn length(&self) -> usize {
         self.length
     }
 
     #[inline]
-    fn new(address: Self::Addr, length: usize) -> Result<Self, CidrParseError> {
+    fn hostmask(&self) -> Self::Addr {
+        if let Some(mask) = u32::max_value().checked_shr(self.length as u32) {
+            Ipv4Addr::from(mask)
+        } else {
+            Ipv4Addr::from(0)
+        }
+    }
+
+    #[inline]
+    fn netmask(&self) -> Self::Addr {
+        Ipv4Addr::from(self.mask)
+    }
+
+    #[inline]
+    fn new(address: Self::Addr, length: usize) -> Result<Self, CidrError> {
         let mask = match length {
             0 => u32::max_value(),
             1..=IPV4ADDR_BITS => u32::max_value() << (IPV4ADDR_BITS - length),
-            _ => return Err(CidrParseError("Not a valid length".to_string())),
+            _ => return Err(CidrError::Malformed("Not a valid length".to_owned())),
         };
 
-        let prefix = u32::from_be_bytes(address.octets()) & mask;
+        let prefix = u32::from(address) & mask;
```

Comment:
same as `length`. we don't use `prefix` much, so I think we can just derive this from mask when needed instead of storing it.